src/ewah/bitset_encoder.zig

```diff
@@ -0,0 +1,234 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const mem = std.mem;
+
+/// `encode`/`decode` compress the bitset using EWAH.
+///
+/// "Histogram-Aware Sorting for Enhanced Word-Aligned Compression in Bitmap Indexes":
+///
+/// > [EWAH] also uses only two types of words, where the first type is a 64-bit verbatim word.
+/// > The second type of word is a marker word: the first bit indicates which clean word will follow,
+/// > half the bits (32 bits) are used to store the number of clean words, and the rest of the bits
+/// > (31 bits) are used to store the number of dirty words following the clean words. EWAH bitmaps
+/// > begin with a marker word.
+///
+/// A 'marker' looks like:
+///
+///     [run_bit:u1][run_length:u31(LE)][literals_count:u32(LE)]
+///
+/// and is immediately followed by `literals_count` 64-bit literals.
+/// When decoding a marker, the run precedes the literals.
```

Comment:
I just mixed up the order in the schema in the documentation, I think everywhere else is correct (including the code & the quote from the whitepaper). It should read:

```
[run_bit:u1][literals_count:u31(LE)][run_length:u32(LE)]
```

I didn't (or at least didn't meant to) imply that the `u32` was unaligned.

---

> Is there another reason I missed (maybe to simplify something?) that you went for this `Marker` struct order: `run_bit, literals, run_length` instead?

I'm not certain I understand your question, but:

- The struct order is so that the `u32` (the `run_length`) is aligned to 4-bytes. This avoids any padding in the packed struct.
- The decision of whether `literals` or `run_length` is `u31` vs `u32` is pretty arbitrary. The whitepaper has `run_length` as the longer field (presumably to optimize for very long runs), and I mimicked that. It doesn't really matter for our use case since our `bitset.length << MAX_U31`.

---

I can change the order and/or the field names if you think that's clearer. But I think `literal`/`run` might be more obvious terminology (notwithstanding the different word lengths). Dirty/clean word is pretty overloaded -- in addition to referring to the literal/run split, it could refer to dirty/clean caches. Or dirty word == profanity, I guess.

Reply:
Agreed on the overloaded meaning for `dirty`/`clean` vs `literal`/`run`... let's keep it clean!

Agreed also that whether `run_length` is 31 or 32 bits is not an issue since 2^31 64-bit words is a huge amount already. I was thinking the same as you also on aligning to 4-bytes, that was what I was trying to convey about not splitting across the half-word boundary.

I do think that `[run_bit:u1][run_length:u31][literals_count:u32]` is a better order then, since it keeps `run_bit`/`run_length` grouped together as we use them, and since it puts the literals after the run, which also matches the actual physical layout (marker, then clean words, then literal words).

As a final try, it would be great to move away from `run` since it's really the mirror image of what we do for `literals` — both are just a count of words with some words being all 0 or all 1 and some words having more entropy than that, somewhere in between 0..1 — so if we can come up with something that makes sense, to have `literals` (plural) and `something else` (plural) and then have `count` for both. But if not, that's fine!

Reply:
Here's one idea running with `extreme` words (either end of the 0/1 spectrum) and `literal` words somewhere in between (don't know if it's good or not!):

```
extreme_value: u1, // Whether the extreme is all 0s or all 1s
extreme_count: u31,
literal_count: u32,
```

Reply:
What about `uniform`?

Reply:
Perfect, looks really clean in the source.